src/main/java/frc/robot/utils/Tracer.java

```diff
@@ -0,0 +1,211 @@
+      gcTimeThisCycle = 0.0;
+
+      // clean up state
+      traceTimes.clear();
+      traceStackHistory.clear();
+    }
+  }
+
+  private static final ThreadLocal<TracerState> threadLocalState =
+      ThreadLocal.withInitial(
+          () -> {
+            return new TracerState(Thread.currentThread().getName());
+          });
+
+  public static void disableGcLoggingForCurrentThread() {
+    TracerState state = threadLocalState.get();
+    state.gcs.clear();
+  }
+
+  private static void startTrace(final String name, final TracerState state) {
+    String stack = state.appendTraceStack(name);
+    TraceStartData data = state.traceStartTimes.get(stack);
+    if (data == null) {
+      data = new TraceStartData();
+      state.traceStartTimes.put(stack, data);
+    }
+    data.set(Logger.getRealTimestamp() / 1000.0, state.totalGCTime());
+  }
+
+  private static void endTrace(final TracerState state) {
+    try {
+      String stack = state.popTraceStack();
+      var startData = state.traceStartTimes.get(stack);
+      double gcTimeSinceStart = state.totalGCTime() - startData.startGCTotalTime;
+      state.gcTimeThisCycle += gcTimeSinceStart;
+      state.traceTimes.put(
+          stack, Logger.getRealTimestamp() / 1000.0 - startData.startTime - gcTimeSinceStart);
+      if (state.traceStack.size() == 0) {
+        state.endCycle();
+      }
+    } catch (Exception e) {
+      DriverStation.reportError(
+          "[Tracer] An end trace was called with no opening trace " + e, true);
+    }
+  }
+
+  /**
+   * Starts a trace, should be called at the beginning of a function thats not being called by user
+   * code. Should be paired with {@link Tracer#endTrace()} at the end of the function.
+   *
+   * <p>Best used in periodic functions in Subsystems and Robot.java.
+   *
+   * @param name the name of the trace, should be unique to the function.
+   */
+  public static void startTrace(String name) {
+    startTrace(name, threadLocalState.get());
+  }
+
+  /**
+   * Ends a trace, should only be called at the end of a function thats not being called by user
+   * code. If a {@link Tracer#startTrace(String)} is not paired with a {@link Tracer#endTrace()}
+   * there could be a crash.
+   */
+  public static void endTrace() {
+    endTrace(threadLocalState.get());
+  }
+
+  /**
+   * Traces a function, should be used in place of {@link Tracer#startTrace(String)} and {@link
+   * Tracer#endTrace()} for functions called by user code like {@code CommandScheduler.run()} and
+   * other expensive functions.
+   *
+   * @param name the name of the trace, should be unique to the function.
+   * @param runnable the function to trace.
+   * @apiNote If you want to return a value then use {@link Tracer#traceFunc(String, Supplier)}.
+   */
+  public static void traceFunc(String name, Runnable runnable) {
+    final TracerState state = threadLocalState.get();
+    startTrace(name, state);
+    runnable.run();
+    endTrace(state);
```

Comment:
This needs a try..finally that ensures endTrace is called otherwise we're going to leak traces on exceptions.

It's also super awkwardly named. I recommend just `trace`.